src/runtime-rs/crates/resource/src/cpu_mem/mem.rs

```diff
@@ -0,0 +1,187 @@
+        let (mem_sb_byte, need_pod_swap, swap_sb_byte) = self
+            .total_mems(self.use_guest_swap)
+            .await
+            .context("failed to calculate total memory requirement for containers")?;
+        let mut mem_sb_mb = (mem_sb_byte >> MIB_TO_BYTES_SHIFT) as u32;
+        mem_sb_mb += self.default_mem;
+        if need_pod_swap {
+            let swap_sb_mb = (swap_sb_byte >> MIB_TO_BYTES_SHIFT) as u32;
+            mem_sb_mb += swap_sb_mb;
+        }
+        info!(sl!(), "calculate mem_sb_mb {}", mem_sb_mb);
+
+        let curr_mem = self
+            .do_update_mem_resource(mem_sb_mb, swap_sb_byte, hypervisor)
+            .await
+            .context("failed to update_mem_resource")?;
+
+        self.update_current_mem(curr_mem).await;
+        Ok(())
+    }
+
+    async fn current_mem(&self) -> u32 {
+        let current_mem = self.current_mem.read().await;
+        *current_mem
+    }
+
+    async fn update_current_mem(&self, new_mem: u32) {
+        let mut current_mem = self.current_mem.write().await;
+        *current_mem = new_mem;
+    }
+
+    async fn total_mems(&self, use_guest_swap: bool) -> Result<(u64, bool, i64)> {
+        // sb stands for sandbox
+        let mut mem_sb = 0;
+        let mut need_pod_swap = false;
+        let mut swap_sb = 0;
+
+        let resources = self.container_mem_resources.read().await;
+
+        for (_, r) in resources.iter() {
+            // Add hugepage memory, hugepage limit is u64
+            // https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L242
+            for l in &r.hugepage_limits {
+                mem_sb += l.limit;
+            }
+
+            if let Some(memory) = &r.memory {
+                let current_limit = match memory.limit {
+                    Some(limit) => {
+                        mem_sb += limit as u64;
+                        info!(sl!(), "memory sb: {}, memory limit: {}", mem_sb, limit);
+                        limit
+                    }
+                    None => 0,
+                };
+
+                // add swap
+                if let Some(swappiness) = memory.swappiness {
+                    if swappiness > 0 && use_guest_swap {
+                        match memory.swap {
+                            Some(swap) => {
+                                if swap > current_limit {
+                                    swap_sb = swap.saturating_sub(current_limit);
+                                }
+                            }
+                            None => {
+                                if current_limit == 0 {
+                                    need_pod_swap = true;
+                                } else {
+                                    swap_sb += current_limit;
+                                }
+                            }
+                        };
+                    }
+                } // end of if let Some(swappiness)
+            } // end of if let Some(memory)
+        }
```

Comment:
some suggestions:
1. split the code block into several functions based on the memory.limit and memory.swapiness;
2. reduce code indentation, enhance code readability.

Reply:
refactored this part